packages/components/src/toolbar/toolbar-item/index.tsx

```diff
@@ -33,14 +42,21 @@ function ToolbarItem( { children, as: Component, ...props }, ref ) {
 		if ( Component ) {
 			return <Component { ...allProps }>{ children }</Component>;
 		}
+		if ( typeof children !== 'function' ) {
+			warning(
+				'`ToolbarItem` is a generic headless component. You must pass either a `children` prop as a function or an `as` prop as a component. ' +
+					'See https://developer.wordpress.org/block-editor/components/toolbar-item/'
+			);
+			return null;
+		}
```

Comment:
If I'm reading the logic correctly, we'll never end up in this conditional block, so I think we can just drop the `warning()` call.

```suggestion
		if ( typeof children !== 'function' ) {
			return null;
		}
```

(Ideally we'd rewrite the logic to be more straightforward so TypeScript can follow better, but I think this is an ok compromise to reduce runtime changes at the moment.)